app/validation/validator.py

```diff
@@ -1207,17 +1206,24 @@ def _validate_placeholder_answer_ids(self, block_id, placeholder_definition, ans
                     placeholder_definition['placeholder']
                 )))
                 continue
+
             answer_block_id = answers_with_parent_ids[answer_id_to_validate]['block']
+
             if answer_block_id == block_id:
                 errors.append(self._error_message('Invalid answer id reference `{}` for placeholder `{}` (self-reference)'.format(
                     answer_id_to_validate,
                     placeholder_definition['placeholder']
                 )))
+                continue
+
+            non_mandatory_error = self._validate_placeholders_reference_mandatory_answers(
+                answers_with_parent_ids[answer_id_to_validate]['answer'])
+            if non_mandatory_error:
+                errors.extend(non_mandatory_error)
```

Comment:
I don't think the extra method is needed. Also the answer id and placeholder name should be in the error message:

```python
answer = answers_with_parent_ids[answer_id_to_validate]['answer'] 
if not answer.get('mandatory'):
    errors.append(self._error_message('Placeholder references a non-mandatory answer `{}` for placeholder `{}`'.format(
        answer_id_to_validate,
        placeholder_definition['placeholder']
    )))
```

Reply:
Are you referring to ` _validate_placeholders_reference_mandatory_answers`?